### ISSUE TITLE
where's scheduler.step? Help implementing SD sonar

### ISSUE THREAD

**Issue by alexblattner** [huggingface/diffusers#4123]
Hi, in the code for SD webui sonar you have this:
```
def sample_euler_ancestral_ex(model:CFGDenoiser, x:Tensor, sigmas:List, extra_args={}, callback=None, eta=1.):
    momentum           = settings['momentum']
    momentum_sign      = settings['momentum_sign']
    momentum_hist      = settings['momentum_hist']
    momentum_hist_init = settings['momentum_hist_init']

    # x: [1, 4, 64, 64], 外源的高斯随机噪声
    show_featmap(x, 'before sample')

    # 记录梯度历史的惯性
    history_d = init_hist_d(momentum_hist_init, x)

    s_in = x.new_ones([x.shape[0]])     # [B=1]
    n_steps = len(sigmas) - 1
    for i in trange(n_steps):
        # [1, 4, 64, 64], 一步降噪后，生成的图越来越明显, sigma理解为步长dx
        denoised = model(x, sigmas[i] * s_in, **extra_args)
        show_featmap(denoised, f'denoised (step {i})')

        # 噪声差分 eps 应当服从正太分布
        eps = denoised - x
        show_featmap(eps, f'eps (step {i})')

        # scalar, sigma_down < sigma_up < sigmas[i + 1] < sigmas[i]
        sigma_down, sigma_up = get_ancestral_step(sigmas[i], sigmas[i + 1], eta=eta)
        if callback is not None: callback({'x': x, 'i': i, 'sigma': sigmas[i], 'sigma_hat': sigmas[i], 'denoised': denoised})
        # # [1, 4, 64, 64], 梯度 = dy/dx = (x - denoised) / sigmas[i]
        d = to_d(x, sigmas[i], denoised)
        show_featmap(d, f'd (step {i}); sigma={sigmas[i]}')

        # ancestral scheduling (down)
        dt = sigma_down - sigmas[i]        # scalar, 沿着梯度移动的步长
        if momentum < 1.0:
            x, history_d = momentum_step(x, d, dt, momentum, momentum_hist, momentum_sign, history_d)
        else:
            x = x + d * dt      # Euler method original
        show_featmap(x, f'x + d x dt (step {i}); sigma_down={sigma_down:.4f}')    # 作画内容逐渐显露
        
        # ancestral scheduling (up)
        x = x + torch.randn_like(x) * sigma_up
        show_featmap(x, f'x + randn (step {i}); sigma_up={sigma_up:.4f}')         # 再被压抑下去
    
    # x: [1, 4, 64, 64], 采样后是否有语义：有，就是生成图的小图，所以vae decoder做的事基本就是超分
    show_featmap(x, 'after sample')

    return x
```

in web ui, the denoising step from what I understand is this: x = x + d * dt

in diffusers, it's this:      latents = self.scheduler.step(result, t, latents, **extra_step_kwargs).prev_sample

how can I manipulate the step function?

**Comment by alexblattner** [comment]
I figured out that it's in the scheduler classes. Is there a way to implement this without modifying to OG schedulers?

**Comment by alexblattner** [comment]
I have finished implementing it for diffusers thanks to @Kahsolt 's help.

for anyone interested: https://github.com/alexblattner/modified-euler-samplers-for-sonar-diffusers
